File: 404/sum_of_left_leaves.go
// 404. Sum of Left Leaves
// https://leetcode.com/problems/sum-of-left-leaves/
package main

import (
	"github.com/k0kubun/pp"
)

func main() {
	pp.Println("=========================================")
	pp.Println()
	pp.Println()
	pp.Println("=========================================")
	pp.Println()
	pp.Println()
	pp.Println("=========================================")
	pp.Println()
	pp.Println()
	pp.Println("=========================================")
	pp.Println()
	pp.Println()
	pp.Println("=========================================")
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sumOfLeftLeaves(root *TreeNode) int {
	var dfs func(*TreeNode, bool) int
	dfs = func(node *TreeNode, isLeft bool) int {
		if node == nil {
			return 0
		}
		if node.Left == nil && node.Right == nil && isLeft {
			return node.Val
		}
		return dfs(node.Left, true) + dfs(node.Right, false)
	}
	return dfs(root, false)
}

File: 404/sum_of_left_leaves.swift
// 404. Sum of Left Leaves
// https://leetcode.com/problems/sum-of-left-leaves/
public class TreeNode {
  public var val: Int
  public var left: TreeNode?
  public var right: TreeNode?
  public init(_ val: Int) {
    self.val = val
    self.left = nil
    self.right = nil
  }
}

class Solution {
  func sumOfLeftLeaves(_ root: TreeNode?) -> Int {
    var dfs: ((TreeNode?, Bool) -> Int)!
    dfs = {(_ node: TreeNode?, _ isLeft: Bool) in
      guard let node = node else { return 0 }
      if node.left == nil && node.right == nil && isLeft {
        return node.val
      }
      return dfs(node.left, true) + dfs(node.right, false)
    }
    return dfs(root, false)
  }
}
